### PR TITLE
Tool checker proposal

### PR DESCRIPTION
**Pull request by denis-yuen**
From https://docs.google.com/document/d/1PTge27WBOKCiR2MkVSOvWKUwnFk0kx-bjq39aKRQxcY/edit?usp=sharing

### CODE REVIEW

**Comment by denis-yuen** [comment]
FYI @briandoconnor  @garyluu

**Comment by briandoconnor** [comment]
Hi @denis-yuen do you think this will be merged and released soon?  By the next GA4GH call?

**Comment by denis-yuen** [comment]
@briandoconnor A few things to be aware of: the current design implies that there is a pointer to the testing workflow for each version of a workflow (although the pointer can be to the same version of the testing workflow). This is also somewhat distinct from the ga4gh-dream challenge which assumed implicitly that a "real" workflow would be run before in the same directory as a testing workflow. 

This allowed the ga4gh-dream participants to avoid writing an separate checker workflow ex: https://github.com/denis-yuen/pcawg-bwa-mem-aligner-unified/blob/master/checker_workflow.cwl which wraps the real workflow and the checker tool. 

This means more work for the workflow author although a "real" workflow like that linked may have an easier time running on workflow platforms that cannot simply reuse a working directory. 

If you're ok with these caveats, please submit a positive review!

**Comment by denis-yuen** [comment]
@jaeddy you might also have thoughts on the consequences of a "real" checker workflow

**Comment by jaeddy** [comment]
I've gone back and forth on how to handle the checker tool vs. workflow question. I think my current preference is what you refer to as a "real" checker workflow comprised of the following steps:

+ run workflow (registered as a workflow) that is being tested
+ run checker tool (registered as a tool) on workflow outputs

Thus, it would make more sense to me to describe the `test-tool-url` field as:
```
test-tool-url:
        type: string
        description: Optional url to the test *tool* that will exit successfully if this workflow produced the expected result given this test data.
```

In the original GA4GH/DREAM challenge, there was never a need for a complete checker workflow, as the steps were split up: participants ran the workflow, I ran the checker tool on the results (however, handling both steps with a single checker workflow on the challenge backend is one of the first things I plan to test out when preparing for the next round). I agree that it would be extra work for workflow authors to write a **third** entry (a checker workflow that wraps the workflow and the checker tool); however, I think there are advantages to not including full workflow execution as an explicit part of the checker tool...

There are a couple ways we could avoid putting the extra burden on authors:

1. Compose checker workflows on the fly: whatever WorkflowCI/orchestrator system we design would know to pull the workflow being tested, look at it's `test-tool-url` field, pull the corresponding checker tool, and chain the steps together in a lightweight and ephemeral descriptor.

2. Create and register a master `check_workflow` workflow that would include the workflow run and checker tool steps, parameterized by JSON inputs pointing to the appropriate descriptor files or URLs.

If you can think of better ways around this, I'm open to being convinced!

**Comment by denis-yuen** [comment]
@jaeddy 
I like 1)  depending on the constraints that we put on the checker tool. There seemed to be a few strategies that people used when I tried upgrading checker tools to checker workflows. 
* some strategies give output files ids and then use the same ids for the inputs to the checker tools with no additional inputs (these would be the easiest to convert)
* some strategies dump everything in the working directory and then feed everything in the current directory back into the checker tool as an array
* some strategies do involve an additional set of reference data, which do complicate things when trying to run a checker workflow with the same parameters as the original workflow (as proposed in the google docs document ... it might be possible to work around this with CWL defaults)

Not sure about 2, is it possible to be that generic with CWL? Might be a question for @mr-c 
(i.e. is it possible to write a generic CWL workflow which runs an arbitrary CWL tool and then a checker tool with the results of that tool? I have some specific examples of a manual process above)

**Comment by briandoconnor** [comment]
@denis-yuen and @jaeddy 

I think the idea of dynamically creating a checker workflow is interesting.  And I like the idea of keeping this as simple as possible for workflow authors.

I'm in the camp of a checker workflow that runs the workflow, then the checker tool, and that checker tool exits with 1|0 depending on if it fails|passes the check.  Hence the overall checker workflow would fail if the check fails.

This is more work initially... I think a workflow author would probably just have a workflow and the checker repo in the same repo.  Now, the reason I want them to go through this extra work on their side is related to non-trivial mappings between workflow outputs and checker tools.  The reality is there might be a complex relationship between the workflow, it's outputs, and the inputs to checker tools.  In fact, a workflow author may have several checker tools that all need to be run.

If we leave this up to Dockstore to map from workflows to checking tools we'll need to give workflow authors a mechanism to specify how to connect the outputs of the workflow to the inputs of the checker tool.  That's quite literally what CWL/WDL do.  Alternatively, we'd end up in a situation where we need to provide that functionality to a developer in Dockstore/TRS and we'd end up replicating the wiring that CWL/WDL provide.

My vote: for a given workflow+version, you optionally specify:
1) test data JSON(s)
2) a workflow for testing (same repo/version tag)

And the workflow author makes sure the JSON test param file(s) can be used either with the workflow or the testing workflow e.g. the contract is they take the same input JSONs to make this maximally easy for CI and also just people running the workflow with "known good" test data.

**Comment by briandoconnor** [comment]
@denis-yuen more questions for you:

In Dockstore will the test workflow be specified as just a path in GitHub along with the main workflow?  This is how workflow paths and test json paths work right now, right?

I'm just wondering about why the endpoints have either URLs or actual content for the test JSON.  And then just a URL for the workflow.  Does it make sense?  Or would it be better to just report them both back as strings (and under the hood the TRS implementation resolves them)?

What do you think?

**Comment by denis-yuen** [comment]
Going a bit in reverse order:

### Why content like descriptors and container recipes can either be URLs or presented as straight content: 

Initially, this was to basically point someone toward the "canonical" source while the content was being returned directly for convenience. If the design of an implementing registry was like Dockstore, the canonical source would be github and we would point there. If the canonical source was the registry itself, it would return the content directly. 
See #21 and #22     
I don't have strong options about this, I think @delagoya actually preferred the URL only in those cases although we went with his alternative which was to make the straight content optional if a URL was provided. 

### Dockstore implementation of test workflows

I think this can be a little independent from what we put in TRS since TRS is supposed to be more general. Either way we do need to decide on whether we require a checker workflow to be written and then whether that is represented as a distinct workflow in Dockstore. (I was previously leaning toward parsing and processing it as a distinct workflow under-the-hood but only "visible" to the user via a path like the test json strategy)

### General strategy 

Basically, I'm undecided. Writing a few of these checker workflows seemed to be a bit of a pain. However, having some arbitrary conventions to code around might be annoying too and might not be portable to different workflow platforms. Both options don't seem great for the workflow developer, so I think we just have to find out which solution minimizes the pain. 

I think we could try to automate generating the skeleton of a checker workflow and a checker tool given a real workflow, so that may help? 

I'm open to ideas here.

**Comment by jaeddy** [comment]
> Now, the reason I want them to go through this extra work on their side is related to non-trivial mappings between workflow outputs and checker tools. The reality is there might be a complex relationship between the workflow, it's outputs, and the inputs to checker tools.

+1 this. Off the top of my head, examples:

1. Workflow outputs are saved in a date/time-stamped directory (see the [`encode_mapping_workflow`](https://www.synapse.org/#!Synapse:syn10163025))
2. Individual workflow output files are date/time-stamped (see the [`pcawg-sanger-variant-caller` workflow](https://www.synapse.org/#!Synapse:syn10517387))
3. Workflow outputs are stored in a directory with unique run ID — which, as far as I could tell, would be the case with _any_ cromwell workflow (see the [`broad-gatk-validate-bam` workflow](https://www.synapse.org/#!Synapse:syn11178580))

There might be other odd cases with directories and relative paths that I'm forgetting. For the GA4GH/DREAM challenge, I typically handled this by modifying the JSON params on the fly — but this was obviously extra work and code on my end, and took some trial and error to understand the connections.

> My vote: for a given workflow+version, you optionally specify:
> 1. test data JSON(s)
> 2. a workflow for testing (same repo/version tag)

So, if I'm understanding correctly, the repo for **my_org/my_workflow::v1.0** would include:
+ `my_workflow.cwl`: main workflow
+ `my_tool.cwl`: used for step in `my_workflow`, source in **my_org/my_tool** repo, image on quay.io
+ `my_workflow_test1.cwl`: test workflow for `my_workflow`
+ `my_workflow_checker1.cwl`: checker tool used in `my_workflow_test1`, source in **my_org/my_workflow_checker1** repo, image on quay.io
+ `my_workflow_test1.cwl.json`: parameters for `my_workflow_test`
+ `known_good1.txt`

> And the workflow author makes sure the JSON test param file(s) can be used either with the workflow or the testing workflow e.g. the contract is they take the same input JSONs to make this maximally easy for CI and also just people running the workflow with "known good" test data.

I'm not sure why this requirement. Assuming that `my_workflow` would never use the `known_good1` input that is needed for `my_workflow_checker1`, why would the test params file be the same? Maybe a better question would be: is there a reason to include test params that would ever be used to _only_ run `my_workflow`, or can we assume that any meaningful testing would be handled through `my_workflow_test1` and thus include the checker tool and its inputs?

**Comment by denis-yuen** [comment]
As a concrete example of my_org/my_workflow there's https://github.com/Barski-lab/ga4gh_challenge/blob/v0.0.2/biowardrobe_chipseq_se_checker.cwl which takes in a control_dir (like `known_good1.txt`) but isn't used in the actual workflow. 

I think the basic appeal of including test params that would ever be used to run my_workflow is for an actual user (as opposed to a CI system). If I want to use a workflow for real work, I'd like to mutate a known good set of parameters to include my data. I think this is why the initial Google Docs proposal targeted the same parameters being used to run both the original workflow and the checker workflow. 
That way, the CI system could test the parameters and the end-user would have a known good to work from. If the two end up being different and only the parameters for the checker workflow are tested, then the end-user would need to reverse-engineer a good set of parameters. 

However, additional parameters like reference files do throw a wrench into things.

**Comment by mr-c** [comment]
I haven't had time to review the entire thread, but I would like to point out that:

 if the outputs of a CWL workflow/tool match the inputs of another CWL workflow/tool
Then:
You can use the output object (the JSON description) as the input object to the other CWL workflow/tool

This allows one to "glue" two CWL descriptions to each other without the need for an encapsulating workflow.

Here is an example: https://github.com/EBI-Metagenomics/ebi-metagenomics-cwl/blob/master/workflows/convert-to-v3-layout.cwl is a follow on to https://github.com/EBI-Metagenomics/ebi-metagenomics-cwl/blob/master/workflows/emg-pipeline-v3-paired.cwl

This should be useful in the context of a workflow and it's checker

**Comment by denis-yuen** [comment]
On one hand, combined with some creative use of defaults to handle things like reference data for checker tools, I think that might work for CWL. 

Unfortunately, I do worry about some of the non-trivial mappings that @jaeddy mentioned above. 

So my thoughts currently are, we might have to go with an explicit checker workflow that ideally has the same inputs (or at least a compatible superset of inputs)  as the original underlying workflow.

**Comment by jaeddy** [comment]
@denis-yuen — can you clarify the difference between `test`, `url` and `test-tool-url` properties for a workflow? I'm having a hard time following the schema.

**Comment by denis-yuen** [comment]
So what we currently have is 
```
 ToolTests:
    type: object
    description: >-
      A tool document that describes how to test with one or more sample test
      JSON.
    properties:
      test:
        type: string
        description: Optional test JSON content for this tool. (Note that one of test and URL are required)
      url:
        type: string
        description: Optional url to the test JSON used to test this tool. Note that this URL should resolve to the raw unwrapped content that would otherwise be available in test.
      test-tool-url:
        type: string
        description: Optional url to the test workflow that will exit successfully if this workflow produced the expected result given this test data. 
```

So one of test or URL is required, both would be great. Test directly holds the JSON parameters for the tool, the URL includes a reference to the JSON parameters for the tool. 

An example probably makes things easier, for example if you wanted to access a Dockerfile, you could go 
```
$ curl -X GET "https://dockstore.org:8443/api/ga4gh/v1/tools/quay.io%2Farhodes_cgrb_osu%2Fdockstore-training-fastqc-ar/versions/latest/dockerfile" -H  "accept: application/json" | python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   856  100   856    0     0   3300      0 --:--:-- --:--:-- --:--:--  3305
{
    "dockerfile": "# FASTQC\nFROM ubuntu:16.04\n\n# File Author / Maintainer\nMAINTAINER Adelaide Rhodes <adelaide.rhodes@gmail.com>\n\n# Install OpenJDK 7 JRE\n#RUN apt-get update && apt-get install --yes \\\n#    default-jre \\\n#    unzip \\\n#    perl-doc \\\n#    perl \n#    docker\n\n# Download FastQC\n#ADD http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.4.zip /tmp/\n\n# Install FastQC\n#RUN cd /usr/local && \\\n#    unzip /tmp/fastqc_*.zip && \\\n#    chmod 755 /usr/local/FastQC/fastqc && \\\n#    ln -s /usr/local/FastQC/fastqc /usr/local/bin/fastqc && \\\n#    rm -rf /tmp/fastqc_*.zip\n\n#ENTRYPOINT [\"fastqc\"]\n\n#ENTRYPOINT [\"/bin/bash\"]\n\nRUN apt-get update && apt-get install -y fastqc perl-doc\nCMD bash\n",
    "url": "https://raw.githubusercontent.com/Arhodes-CGRB-OSU/dockstore-training-fastqc-ar/master/Dockerfile"
}
```

dockerfile contains the Dockerfile, url contains where we got it from. 

Similarly, test contains the test parameters, url contains where we got it from. 
The proposal also adds test-tool-url which is a URL to another workflow in the TRS which is a checker workflow.

**Comment by jaeddy** [comment]
Thanks, Denis. Mapping to my example from above, would this be correct?

+ `Tool` ~ `my_workflow.cwl` (main workflow)
+ `test` ~ `test.cwl.json` (simple test of tool/workflow, also serves as example of parameters for tool)
+ `url` ~ `<url-path-to>test.cwl.json`
+ `test-tool-url` ~ `<url-path-to>my_workflow_test1.cwl` (checker workflow for `my_workflow`)

What's not explicitly covered in the spec is therefore:
+ `my_workflow_checker1.cwl` (checker tool used by `my_workflow_test1`)
+ `my_workflow_test1.cwl.json`(parameters for `my_workflow_test1`; though maybe this would be part of the entry for `my_workflow_test1`?)

(sorry if I'm misunderstanding still)

**Comment by denis-yuen** [comment]
No worries, 
The latter two would be part of the entry for `my_workflow_test1`. Since `my_workflow_test1` would be a full entry in the TRS, it would get its own associated secondary files like `my_workflow_test1` and its own json parameters like `my_workflow_test1.cwl.json` which may or may not be the same as the json parameters as the originals.

**Comment by denis-yuen** [comment]
FYI, squashing this into `develop` but feel free to continue the discussion in issues. There's clearly a lot to discuss
